be/src/exprs/bitmap_function.cpp

```diff
@@ -22,6 +22,256 @@
+        if (it != _keys.end()) {
+            if (_bitmaps.find(key) != _bitmaps.end()) {
+                _bitmaps[key].merge(bitmap);
+            } else {
+                _bitmaps[key] = bitmap;
+            }
+        }
+    }
+
+    void merge(const BitmapRetention& other) {
+        // Note, in merge phase, this _keys is empty and other's _keys is whole
+        if (_keys.empty()) {
+            _keys = other._keys;
+        }
+        for(auto& key: _keys) {
+            if (other._bitmaps.find(key) == other._bitmaps.end()) {
+                continue;
+            }
+
+            if (_bitmaps.find(key) != _bitmaps.end()) {
+                _bitmaps[key].merge(other._bitmaps.at(key));
+            } else {
+                _bitmaps[key] = other._bitmaps.at(key);
+            }
+        }
+
+    }
+
+    int64_t result() {
+        if (_keys.empty()) {
+            return 0;
+        }
+
+        // if any specified key not in map, the intersection must be 0
+        for(auto& key: _keys) {
+            if (_bitmaps.find(key) == _bitmaps.end()) {
+                return 0;
+            }
+        }
+
+        for(int32_t i = 1; i < _keys.size(); ++i) {
+            _bitmaps[_keys[0]].intersect(_bitmaps[_keys[i]]);
+        }
+
+        return _bitmaps[_keys[0]].cardinality();
+    }
+
+    size_t size() {
+        size_t size = 4; // keys size;
+        for (auto& key: _keys) {
+            size += detail::serialize_size(key);
+        }
+
+        size += 4; // bitmaps size;
+        for (auto& kv: _bitmaps) {
+            size +=  detail::serialize_size(kv.first);;
+            size +=  kv.second.size();
+        }
+        return size;
+    }
+
+    //must call size() first
+    void serialize(char* dest) {
+        char* writer = dest;
+        *(int32_t*)writer = _keys.size();
+        writer += 4;
+        for (auto& key: _keys) {
+            writer = detail::write_to(key, writer);
+        }
+
+        *(int32_t*)writer = _bitmaps.size();
+        writer += 4;
+        for (auto& kv: _bitmaps) {
+            writer = detail::write_to(kv.first, writer);
+            kv.second.serialize(writer);
+            writer += kv.second.size();
+        }
+    }
+
+    void deserialize(const char* src) {
+        const char* reader = src;
+        int32_t keys_size = *(int32_t*)reader;
+        reader += 4;
+        for(int32_t i = 0; i < keys_size; i++) {
+            T value;
+            detail::read_from(&reader, &value);
+            _keys.push_back(value);
+        }
+
+        int32_t bitmaps_size = *(int32_t*)reader;
+        reader += 4;
+        for(int32_t i = 0; i < bitmaps_size; i++) {
+            T key;
+            detail::read_from(&reader, &key);
+            RoaringBitmap bitmap(reader);
+            reader += bitmap.size();
+            _bitmaps[key] = bitmap;
+        }
+    }
+
+private:
+    std::map<T, RoaringBitmap> _bitmaps;
+    std::vector<T> _keys;
```

Comment:
Is it possible to only keep the map of bitmaps? When key is added, bitmap will be put in the map

Reply:
The `_keys ` means the constant args. it's a super set of bitmap keys.  I will rename it

Reply:
What I want to emphasize is that assign a empty bitmap for each constand arg. Change it from allocating on demand to allocating ahead.